File: src/apps/daemon/internal/runtime/proxy/static_cidr_pool.go
package proxy

import (
	"errors"
	"net"
	"sync"
)

type StaticClientRecord struct {
	ClientID   string `json:"client_id"`
	AllocatedIP net.IP `json:"allocated_ip"`
	PublicKey  string `json:"public_key"`
	IsRevoked  bool   `json:"is_revoked"`
}

type StaticCidrPool struct {
	mu          sync.Mutex
	basePrefix  [3]byte
	nextHost    byte
	allocated   map[string]*StaticClientRecord
	revokedKeys map[string]bool
}

func NewStaticCidrPool(prefix [3]byte) *StaticCidrPool {
	return &StaticCidrPool{
		basePrefix:  prefix,
		nextHost:    2,
		allocated:   make(map[string]*StaticClientRecord),
		revokedKeys: make(map[string]bool),
	}
}

func (p *StaticCidrPool) AllocateClient(clientID, publicKey string) (*StaticClientRecord, error) {
	p.mu.Lock()
	defer p.mu.Unlock()

	if p.nextHost >= 254 {
		return nil, errors.New("cidr pool exhausted")
	}

	ip := net.IPv4(p.basePrefix[0], p.basePrefix[1], p.basePrefix[2], p.nextHost)
	p.nextHost++

	rec := &StaticClientRecord{
		ClientID:    clientID,
		AllocatedIP: ip,
		PublicKey:   publicKey,
		IsRevoked:   false,
	}
	p.allocated[ip.String()] = rec
	return rec, nil
}

func (p *StaticCidrPool) RevokeClient(ip net.IP) bool {
	p.mu.Lock()
	defer p.mu.Unlock()

	if rec, exists := p.allocated[ip.String()]; exists {
		rec.IsRevoked = true
		p.revokedKeys[rec.PublicKey] = true
		return true
	}
	return false
}

func (p *StaticCidrPool) IsKeyRevoked(pubKey string) bool {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.revokedKeys[pubKey]
}

File: src/packages/lumicore/src/wg/static_cidr_pool.rs
use std::collections::{HashMap, HashSet};
use std::net::Ipv4Addr;

#[derive(Debug, Clone, PartialEq)]
pub struct AllocatedClient {
    pub client_id: String,
    pub ip: Ipv4Addr,
    pub public_key: String,
    pub is_revoked: bool,
}

#[derive(Debug)]
pub struct StaticCidrPool {
    base_prefix: [u8; 3],
    allocated: HashMap<Ipv4Addr, AllocatedClient>,
    revoked_keys: HashSet<String>,
    next_host: u8,
}

impl StaticCidrPool {
    pub fn new(prefix: [u8; 3]) -> Self {
        Self {
            base_prefix: prefix,
            allocated: HashMap::new(),
            revoked_keys: HashSet::new(),
            next_host: 2, // 1 is gateway
        }
    }

    pub fn allocate_client(&mut self, client_id: &str, public_key: &str) -> Option<AllocatedClient> {
        if self.next_host >= 254 {
            return None;
        }
        let ip = Ipv4Addr::new(self.base_prefix[0], self.base_prefix[1], self.base_prefix[2], self.next_host);
        self.next_host += 1;

        let client = AllocatedClient {
            client_id: client_id.to_string(),
            ip,
            public_key: public_key.to_string(),
            is_revoked: false,
        };
        self.allocated.insert(ip, client.clone());
        Some(client)
    }

    pub fn revoke_client(&mut self, ip: &Ipv4Addr) -> bool {
        if let Some(client) = self.allocated.get_mut(ip) {
            client.is_revoked = true;
            self.revoked_keys.insert(client.public_key.clone());
            true
        } else {
            false
        }
    }

    pub fn is_key_revoked(&self, public_key: &str) -> bool {
        self.revoked_keys.contains(public_key)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_static_cidr_allocation_and_revocation() {
        let mut pool = StaticCidrPool::new([10, 66, 0]);
        let client = pool.allocate_client("user1", "pubkeyA111=").expect("should allocate");

        assert_eq!(client.ip, Ipv4Addr::new(10, 66, 0, 2));
        assert!(!pool.is_key_revoked("pubkeyA111="));

        assert!(pool.revoke_client(&client.ip));
        assert!(pool.is_key_revoked("pubkeyA111="));
    }
}
